blueprints/multi-interface-servers.md

```diff
@@ -0,0 +1,666 @@
+
+If desired, the markup used for that sample may be found in
+[Appendix A](#appendix-a)
+
+Additionally, the properties of servers no longer present on a server object
+will need to be removed from table columns (and `interfaces` should not be
+added).
+
+### Traffic Monitor Impact
+Traffic Monitor will need to update its parsing of `/monitoring` payloads to
+account for the new structure, and will need to poll all listed interfaces for
+health (instead of just one per cache server as was done previously).
+
+The cache
+server should be marked unavailable - or "down" - if and only if the interface
+proving the service addresses is considered unhealthy.
+
+## Documentation Impact
+Documentation for the affected endpoints will need to be updated. Beyond that,
+the "Health Protocol" section and perhaps the "Traffic Monitor - Using" section
+should be updated to describe the behavior of multiple monitored interfaces.
+
+## Testing Impact
+A bunch of client/API integration tests should be written to verify the
+constraints outlined in [API](#sec:api).
+
+## Performance Impact
+The proposed method of updating servers to change their interfaces is by
+submitting a PUT request to `/servers/{{ID}}` like normal. Traffic Ops will then
+delete **all** associated interfaces before inserting the ones supplied in the
+PUT payload. This will occur even if the interfaces are not actually changing in
+values. So a single server update now requires a minimum of two extra database
+calls.
+
+This probably isn't a significant performance concern, but at the time of this
+writing no profiling has been done.
+
+## Security Impact
+Setting an improper interface can allow users with read-only permissions to read
+arbitrary files named 'speed' or 'operstate' that the 'ats' user has permission
+to read on the cache server's file system. Of course, this was already possible,
+and improper values can only be entered by users with at least "operator"
+permissions. Now, though, an arbitrary number of these files could be exposed
+per Snapshot, whereas previously it was limited to one file per cache server per
+Snapshot.
+
+Improper configuration has the same result as improper configuration of
+`interfaceName` currently, which is to say it can cause a cache server to be
+marked unavailable.
+
+Multiple new inputs will need to be sanitized, as always.
+
+> _"Will new SQL queries properly use parameter binding?"_
+
+The answer to this question should absolutely never be "no," can't imagine a
+world wherein that would need to change for this - or any other - particular
+feature.
+
+## Upgrade Impact
+There will be at least one, but possibly two, database migrations. The changes
+to Traffic Monitor ought to be made in such a way that it can parse either the
+old `/monitoring` payloads or the new ones. Then Traffic Monitor should be
+upgraded before Traffic Ops, and no additional, special steps need be taken to
+upgrade.
```

Comment:
What's with the deployment order dependency here? Are we anticipating breaking the old interfaces such that we have to deploy TM first?  I usually think that we want to deploy TO first so I want to be sure to understand why that would change.

Reply:
Well, the way this will work is this: if these changes get into the 2.0 API release, you can upgrade TO first because TM will still be using the 1.x API.

And actually doing anything else would be a violation of semantic versioning, because we're removing fields not just adding them. So if this doesn't make it into API 2.0 it'd need to go in 3.0.

Reply:
OK, so if we bump the API to 3.0 with this new API then we don't have an issue, right?

Reply:
Right